src/streamlink_cli/main.py

```diff
@@ -130,11 +132,35 @@ def create_http_server(host=None, port=0):
 
 def create_title(plugin=None):
     if args.title and plugin:
-        _title = plugin.get_title()
-        _author = plugin.get_author()
-        _category = plugin.get_category()
-        _game = _category
+        _title = _author = _category = _game = None
+        
+        #"did the user include a format code?" logic
+        tokens = ["t17L3_XXQZ","4uTh0r_qZ","c4Teg0ry_zQ","G4m3_z4"]
```

Comment:
What's this trying to do, looks very cryptic to me :)

Reply:
basically just "create a substring not in original string" so that we can use the same logic of string.format to see if we will alter the `--title` string given by the user without creating key errors. 

we can't just do `if args.title != args.title.format(title="different")` because if `{author}` is in args.title then we get a key error. So I create these tokens that are defined in a way that they cannot possibly be substrings of `args.title`, and we're still able to track which format code was replaced.

If those substring tokens appear in arg_title_formatted, then I know that args.title.format put them there. and I can compare what token is in args_title_formatted to determine which format code was in the original string.

a simpler implementation would be something like `if "{author}" in args.title:`, but "what if the user wants to put {{author}} in the title?". I had flashbacks to when I implemented custom quotes in #623 and everyone hated it. This implementation uses no regex's and only uses the same exact string.format logic to detect if string.format will string.format the string.

Reply:
> So I create these tokens that are defined in a way that they cannot possibly be substrings of args.title

This method was one of the reasons why the previous PR was rejected. I haven't looked into this PR yet (was just briefly reading the comments), but doing string replacement based on (static) tokens which are "not likely to be used" is bad. If this method really has to be used here, at least generate a random prefix with an additional timestamp.

Reply:
it is literally impossible now because we loop until we generate string not in args.title.

Reply:
So the idea is you want to avoid calling the `get_{author,title,game}` method unless the format code is in the title?

Reply:
correct, that's what this does. it prevents unnecessary api calls.

Reply:
Here is an alternative, more generic (but no less magical), implementation... 

https://gist.github.com/beardypig/14da8dc260cb249b3c0d7508130b1d24

Reply:
This looks better. :) I'll put `class LazyFormatter(object):` into `streamlink_cli/util/lazy_formatter.py` and use it.

Reply:
Needs a couple more tests for non-callable items.

Reply:
`create_title` currently looks like this if you're writing tests for `LazyFormatter`. I had to change it so that it could supply a default value.
```
def create_title(plugin=None):
    if args.title and plugin:
        _title = _author = _category = _game = None
        
        title = LazyFormatter.format(maybe_encode(args.title),
                                     title=[plugin.get_title(),DEFAULT_STREAM_METADATA["title"]],
                                     author=[plugin.get_author(),DEFAULT_STREAM_METADATA["author"]],
                                     category=[plugin.get_category(),DEFAULT_STREAM_METADATA["category"]],
                                     game=[plugin.get_category(),DEFAULT_STREAM_METADATA["game"]])
    else:
        title = args.url
    return title
```

Reply:
Unfortunately this defeats the purpose of the `LazyFormatter`. 
What you have now is the same as 
```python
maybe_encode(args.title).format(
  title=plugin.get_title() or DEFAULT_STREAM_METADATA["title"],
  author=plugin.get_author() or DEFAULT_STREAM_METADATA["author"],
  ...
)
```
ie. The methods always get called.

Reply:
Hmm, okay. I did notice that just putting each function in an array seemed to call it.

DEFAULT_STREAM_METADATA must be available at the time of formatting the final title, so I'll import it inside LazyFormatter instead of supplying it here.

Reply:
To have a default value use the original `LazyFormatter` I proposed,

```python
title = LazyFormatter.format(maybe_encode(args.title),
                             title=lambda: plugin.get_title() or DEFAULT_STREAM_METADATA["title"],
                             author=lambda: plugin.get_author() or DEFAULT_STREAM_METADATA["author"],
                             category=lambda: plugin.get_category() or DEFAULT_STREAM_METADATA["category"],
                             game=lambda: plugin.get_category() or DEFAULT_STREAM_METADATA["game"])
```

That way the methods are only called when needed.

Reply:
ok, thanks. :)

Reply:
I have updated the `LazyFormatter` slightly, it not allows the format string to contain `{fmt}`. 

https://gist.github.com/beardypig/14da8dc260cb249b3c0d7508130b1d24

All the tests are included too, so you can include that file in the `tests` directory and just move `LazyFormatter` out to utils.